ourhexenv.py

```diff
@@ -171,25 +172,31 @@ def reset(self):
             pygame.display.flip()
 
     def step(self, action):
-        #Check if the action is within the valid range.
+    # Check if the action is within the valid range
         if action not in range(self.action_space.n):
             raise ValueError("Illegal move: Action is out of bounds.")
-        # Handle pie rule
+
+        # Handle pie rule (only Player 2 can use it in the first round)
         if action == self.board_size * self.board_size:
             if self.agent_selection == "player_1":
                 raise ValueError("Illegal move: Pie rule can only be used by Player 2.")
             if not self.is_pie_rule_usable:
                 raise ValueError("Illegal move: Pie rule can only be used once.")
+            if np.sum(self.board == 1) == 0:
+                raise ValueError("Illegal move: Pie rule cannot be used before Player 1's move.")
 
-            # Use pie rule, if a (row,col) was 1, make it 0 and make (col,row) 1
+            # Flip the board state 
             x, y = np.where(self.board == 1)
             row, col = x[0], y[0]
             self.board[row][col] = 0
             self.board[col][row] = 2
+
+            # Update flags
+            self.pie_rule_used = True
+            self.is_pie_rule_usable = False
```

Comment:
This can be removed. This logic is there in line #249

```
if self.agent_selection == "player_2":
    self.is_pie_rule_usable = False
```

Reply:
In general, keeping a side effect close to the code that triggered it; is better. In my opinion, it makes more sense to **keep** this as you can easily understand that the pie rule triggered this side effect.

Reply:
After re-reading this, adding this **only** here will be a bug because it only invalidates the flag if the pie rule was used. However, the is_pie_rule_usable should be invalidated based on the turn, not on whether the pie rule was passed or not. I agree with you, remove the line
`self.is_pie_rule_usable = False`
 
and add it closer  to the end of the step method, as you had